app/controllers/application_controller.rb

```diff
@@ -929,6 +930,29 @@ def should_enforce_2fa?
       !current_user.has_any_second_factor_methods_enabled?
   end
 
+  def redirect_to_profile_if_required
+    return if request.format.json? && is_api?
+    return if !current_user
+    return if !current_user.needs_required_fields_check?
+
+    if current_user.populated_required_custom_fields?
+      current_user.bump_required_fields_version
+      return
+    end
+
+    redirect_path = path("/u/#{current_user.encoded_username}/preferences/profile")
+    second_factor_path = path("/u/#{current_user.encoded_username}/preferences/second-factor")
+    allowed_paths = [redirect_path, second_factor_path, "/admin"]
+    if allowed_paths.none? { |p| request.fullpath.start_with?(p) }
+      UserHistory.create!(
```

Comment:
This only needs to be once per day, otherwise the user could end up making tons of these. This can be done in redis with a combination of `setnx` and `expire`, here is another example:

https://github.com/discourse/discourse/blob/a1e330e6504988ed24d4b792215b717f4484c34a/app/models/topic_link_click.rb#L109-L114

If `setnx` returns true, it means the value did not exist https://github.com/redis/redis-rb/blob/433ec69c563f1b51dead336fc81f1fb70a8297af/lib/redis/commands/strings.rb#L125

It would be nice to eventually extract this pattern into a useful oneliner, like:

```ruby
Discourse.redis.perform_once_in_period(key, seconds) do
  // the thing
end
```

Reply:
This and the 302 error on redirect from /admin back to the forum are the most important things to change before merging

Reply:
Can you also please write a spec for it

Reply:
> It would be nice to eventually extract this pattern into a useful oneliner.

I think this could even go in as part of the `UserHistory` API in addition to having the low-level one-liner.